pkg/infrastructure/openstack/clusterapi/clusterapi.go

```diff
@@ -31,6 +31,10 @@ func (p Provider) Name() string {
 	return openstack.Name
 }
 
+// BootstrapHasPublicIP indicates that an ExternalIP is not
+// required in the machine ready checks.
+func (Provider) BootstrapHasPublicIP() bool { return false }
```

Comment:
In OpenStack, we add a Floating IP to the bootstrap node, but we do it inside the `PostProvision` hook, meaning that, with the current implementation, we don't want to set `BootstrapHasPublicIP()` to return `true`, otherwise we're entering a deadlock where `Provision()` waits for the machine to have a FIP while the FIP is only added later.

On a related note, I reported https://issues.redhat.com/browse/OCPBUGS-34950 yesterday while working on a different issue. It looks to me that we want to write the manifests once again after `PostProvision` so that it has the external IP for the bootstrap?

Reply:
@patrickdillon maybe we can move the IP waiting to the PostProvision hook?

Reply:
> we don't want to set BootstrapHasPublicIP() to return true, otherwise we're entering a deadlock where Provision() waits for the machine to have a FIP while the FIP is only added later. otherwise we're entering a deadlock where Provision() waits for the machine to have a FIP while the FIP is only added later.

Right. It is set to `false` here in order to avoid a deadlock.

> /hold
> for the possible deadlock in openstack.

@mandre can you clarify whether the deadlock you mentioned was potential/hypothetical? As I understand this PR will not currently deadlock.

> On a related note, I reported https://issues.redhat.com/browse/OCPBUGS-34950 yesterday while working on a different issue. It looks to me that we want to write the manifests once again after PostProvision so that it has the external IP for the bootstrap?

The manifests are written after PostProvision, so a solution like the one proposed in https://github.com/openshift/installer/pull/8492 would work. 

> @patrickdillon maybe we can move the IP waiting to the PostProvision hook?

I'm open to the idea, but I don't think we don't need to. #8492 should handle it

Reply:
> I'm open to the idea, but I don't think we don't need to. https://github.com/openshift/installer/pull/8492 should handle it

It's probably a bad idea since if we fail during Provisioning we are back to not having logs if the machine has no IP :sweat_smile:

Reply:
Just to be clear, with the current implementation where `BootstrapHasPublicIP()` returns `false` there's no deadlock. We'll try to sort it out in #8492.